src/server/pfs/server/obj_block_api_server.go

```diff
@@ -322,7 +321,7 @@ func (s *objBlockAPIServer) GetObject(request *pfsclient.Object, getObjectServer
 		return err
 	}
 	objectSize := objectInfo.BlockRef.Range.Upper - objectInfo.BlockRef.Range.Lower
-	if (objectSize) >= uint64(s.objectCacheBytes/maxCachedObjectDenom) {
+	if s.objectCacheBytes < int64(objectSize) || int64(objectSize) > pfsclient.ChunkSize {
```

Comment:
So it seems like this PR is substantially lowering the object size at which we bypass the cache from `1/4` of total cache size which I think was 250MB by default to 16 MB by default. It seems like using a cache for objects greater than 16MB in size might make a lot of sense. What's the reasoning behind this change?

Reply:
The idea is that most objects will be <= 16MB due to how `PutFile` works today.  So most objects should be cache-able.  The problem was that even though a sidecar had 64MB of cache by default, we are bypassing the cache for 16MB objects since `16 == 64/4`, which is not ideal, since we definitely do want to cache those objects.  So I'm changing the code so that the cache is only bypassed for objects larger than 16MB.  In reality, there's only one way for an object to be greater than 16MB, which is to have a huge serialized hashtree.

Reply:
This seems very sub optimal to me. The whole point of this code was to prevent us from blowing our entire cache on a large object. That can now happen easily. Suppose we have an object that's the same size as our cache. A `get-file` for that will cause the following to happen:

- read hashtree into cache
- read all chunks for object into cache (the last one of which will evict the hashtree)

If someone reads the same file again we'll have to start by reading in the hashtree, which will evict the first chunk, then we read the first chunk which evicts the second chunk, etc. So the cache is completely useless for this workload, and useless for all other workloads while it's running.

> The problem was that even though a sidecar had 64MB of cache by default, we are bypassing the cache for 16MB objects since 16 == 64/4, which is not ideal, since we definitely do want to cache those objects.

Do we definitely want to cache those objects? If we're caching those objects then it's only going to be useful if there's <3 of them, since we need some space for the hashtree as well. If there's more than 3 of them then we'll like wind up in a situation like the above where we never read things from the cache, since they're constantly getting evicted by other big objects.

I think we should go back to the old behavior i.e. bypass if it's greater than 1/4 the size of the cache. However we should tweak it so that with `GetObjects` we either cache all the objects or don't based on the cumulative size of the objects (and whether they're more than 1/4) of the cache. It doesn't really make sense to decide for each individual object since they're almost certainly all going to be read at the same time.

Reply:
The point about how a large object would render the cache useless seems to make sense.  Agree that we should consider the cumulative size with regard to eviction.  I will change it.

Reply:
:+1: